### ISSUE TITLE
Hide and Unhide Columns and Sheets?

### ISSUE THREAD

**Issue by Patowhiz** [IDEMSInternational/R-Instat#4372]
Well, probably this was discussed before and could be clarified to me. Currently to hide a column, I have to go to Prepare > Data Frame > Hide Columns . Could we have a short cut for this in when I right click on column, probably just after the duplicate column. I find hide column as one of the most frequent functionalities used in grid like user controls. Under this shortcut, the dialog doesn't necessarily have to appear, it could just call the function involved in hiding column directly.

Then on the Prepare > Data Frame > Hide Columns which is better between that and Prepare > Data Frame > Hide/Show Columns . I believe the dialog does both functions.

**Comment by dannyparsons** [comment]
We actually did have this on the right click menu originally. Then the menu got very long and we removed lots of items, including the 2 hide items. Happy to consider adding it back now if we feel it's needed. We would need 2 items, 1 to just hide the selected columns and 1 to go to the dialog to be able to unhide columns. @rdstern any thoughts?

**Comment by Patowhiz** [comment]
@rdstern what are you thoughts about this. I can close this issue if this is not desirable. In addition to the above, I noticed the menu item is labeled **Hide column**. Is this preferable than **Hide/show column**. I find the latter much more descriptive.

**Comment by rdstern** [comment]
First the small one.  Yes I think the menu item should be **Hide/Show Columns**.  I had forgotten this feature, perhaps because it was not on the right/click menu.

We could add it back, but there may be complications?
However, I do like some of its current features already.
In a "narrow" data set you can see that some columns are hidden, because it says "Showing 9 out of 12 columns".  That's good.
In a wide data set (158 columns) it still says "Showing 30 out of 158 columns." when I hide 6 columns.  That's good and bad.  It is good because it is showing 6 more new columns and I like that.  It is bad because it is not clear now that any columns are hidden.  

So when columns are hidden, then do we want to add "(6 hidden)", i.e. "Showing 30 of 158 columns (6 hidden)"?
We also have the same menu when you right-click on the meta-data.  It doesn't apply there, so would it be easy to not have those items then?  

I am keen to have the items on right-click that would be useful and I can see this could be one of them.  I would only want this change if it is quick, because there is so much else to do just now.

And it ties (for me anyway) with the items we have left there, but that are disabled, namely the freeze columns.  They seem to be partners.  Could we consider the issues quickly together?

**Comment by Patowhiz** [comment]
@rdstern this can be done quickly.
I think "Showing 30 of 158 columns (6 hidden)" would be correct for a wide datasets.
Should we add the hide column item before Delete Column in the context menu ?

<img width="255" alt="column reorder" src="https://user-images.githubusercontent.com/21179007/40107622-695309c8-5901-11e8-8dcb-ff2290fc05f9.png">

**Comment by dannyparsons** [comment]
Before we had Hide and Unhide as two items on this menu, and then removed them, maybe because the menu was getting too long, can't remember. So if we add back hide, we need to be able to unhide from the right click as well (which goes to the Hide dialog). 
@Patowhiz what was the change done in #4602?

**Comment by Patowhiz** [comment]
@dannyparsons pull request #4602 had changes in frmMain where I changed the menu item form just hide column to hide/show column. Maxwell suggested we merge it first since it touched frmMain.

**Comment by Patowhiz** [comment]
@dannyparsons I don't find it logical to have unhide on the right click because right click touches on the visible column only. Any action made affects that visible column only. I think having hide, and making it do the action immediately instead of going to the dialog will be better from the user's perspective

**Comment by rdstern** [comment]
I have an alternative suggestion.  I also feel we should be careful if we are close to, but different from Excel.
I am not sure what Hide is doing currently in R-Instat?  Is it automatically hiding the selected columns?  Or does it go to the dialogue.  I am quite happy if it hides without going to the dialogue.

But then there is the dialogue which is more general than the right-click, because we can't select non-adjacent columns in R-Instat.  

So, now to the Unhide.  If that is on our usual Right-click, then the only simple action is to unhide all hidden columns.  That isn't then the exact opposite of Hide.

An alternative is to argue that Unhide Columns is an action for the Sheet as a whole, so logically it could go on the bottom tab instead.  If so, then we should call it Unhide Columns, because I presume we should also want to Hide and Unhide sheets.  Is that easy in the grid?  That is what Hide and Unhide do in Excel.  So we would then have Hide Sheet and Unhide Sheet also on the bottom tab.  And we have plenty of space there.

**Comment by Patowhiz** [comment]
@rdstern the hide column in the Prepare>Data Frame menu goes to the dialog. That's why I would prefer the right click hide column to hide the column directly(It seems logical cause the column is already selected and the experience will be similar to other softwares).

I second you in the unhide suggestion. Having it at the bottom is a viable option. The Hide Sheet and unhide sheet are already on the bottom tab on right click, so it's just a matter of adding the unhide columns which when clicked goes to the hide columns dialog.

**Comment by Patowhiz** [comment]
@dannyparsons based on your opinion I could start this right away.

**Comment by rdstern** [comment]
Yes, I had just re-discovered that we already have hide and unhide sheet.  If we do add Unhide Columns there, then perhaps change Hide to Hide Sheet (and add dot,dot, dot!).  And Unhide becomes Unhide Sheet.

**Comment by rdstern** [comment]
It is illogical that we have Hide and Unhide on the Sheets right-click and not on the Columns right-click.
I suggest:
a) Add Hide/Show... to the Column right-click.  It is a single entry and always visible and goes to the Prepare > Data Frame Hide/Show Columns Dialogue.
b) Change the 2 entries for the Sheet Hide and Unhide - I think they are set wrong now.  Again they are Hide/Show... and they go to the Prepare > Data Object > Hide/Show Data Frames dialogue.
c) When Columns are hidden this is stated in the text at the bottom (x columns hidden) is at the end of the description.  
d) I don't know what to do to show there are hidden data frames.

**Comment by shadrackkibet** [comment]
@africanmathsinitiative/developers the last bullet point remains here. Any volunteer to try it out? Very happy to give guidance on how to do it.
